src/Renci.SshNet/Session.cs

```diff
@@ -650,6 +650,11 @@ public void Connect()
                     // ToDo: Make message pump async, to not consume a thread for every session
                     _ = ThreadAbstraction.ExecuteThreadLongRunning(MessageListener);
 
+                    if (ConnectionInfo.InitiateKeyExchange)
+                    {
+                        SendMessage(ClientInitMessage);
+                    }
```

Comment:
So to summarise my thoughts here, we can

1. make this unconditional
2. add `_keyExchangeInProgress = true` before the `SendMessage(ClientInitMessage);` call
3. move it above line 651 (`_ = ThreadAbstraction.ExecuteThreadLongRunning(MessageListener);`). This will ensure we have sent the client init before we continue the key exchange
4. Add a parameter `bool sendClientInitMessage` to `IKeyExchange.Start` which guards `SendMessage(session.ClientInitMessage);` in `KeyExchange.Start`
5. Add a `private bool _initialKeyExchangeCompleted;` (or similar) member to `Session`, which is used to decide the `sendClientInitMessage` parameter in `Session.OnKeyExchangeInitReceived`.

As for testing, if we make it unconditional then we are covered already 🙂